File: pulsar-skeleton/src/main/java/ai/platon/pulsar/crawl/fetch/FetchThread.java
package ai.platon.pulsar.crawl.fetch;

import ai.platon.pulsar.common.MetricsSystem;
import ai.platon.pulsar.common.ReducerContext;
import ai.platon.pulsar.common.StringUtil;
import ai.platon.pulsar.common.Urls;
import ai.platon.pulsar.common.config.ImmutableConfig;
import ai.platon.pulsar.crawl.component.FetchComponent;
import ai.platon.pulsar.crawl.fetch.data.PoolId;
import ai.platon.pulsar.persist.WebPage;
import ai.platon.pulsar.persist.metadata.FetchMode;
import org.slf4j.Logger;

import javax.annotation.Nonnull;
import java.io.IOException;
import java.net.URL;
import java.util.Objects;
import java.util.Set;
import java.util.TreeSet;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Collectors;

import static ai.platon.pulsar.common.config.CapabilityTypes.FETCH_MODE;

/**
 * This class picks items from queues and fetches the pages.
 */
public class FetchThread extends Thread implements Comparable<FetchThread> {

    private final Logger LOG = FetchMonitor.LOG;
    private final Logger REPORT_LOG = MetricsSystem.REPORT_LOG;

    private static AtomicInteger instanceSequence = new AtomicInteger(0);

    private ReducerContext context;

    private final int id;

    private final FetchComponent fetchComponent;
    private final FetchMonitor fetchMonitor;
    private final TaskScheduler taskScheduler;
    /**
     * Native, Crowdsourcing, Proxy
     */
    private final FetchMode fetchMode;

    /**
     * Fix the thread to a specified queue as possible as we can
     */
    private int currPriority = -1;
    private PoolId currQueueId = null;
    private AtomicBoolean halted = new AtomicBoolean(false);
    private Set<PoolId> servedHosts = new TreeSet<>();
    private int taskCount = 0;

    public FetchThread(
            FetchComponent fetchComponent,
            FetchMonitor fetchMonitor,
            TaskScheduler taskScheduler,
            ImmutableConfig immutableConfig,
            ReducerContext context) {
        this.context = context;

        this.fetchComponent = fetchComponent;
        this.fetchMonitor = fetchMonitor;
        this.taskScheduler = taskScheduler;

        this.id = instanceSequence.incrementAndGet();

        this.setDaemon(true);
        this.setName(getClass().getSimpleName() + "-" + id);

        this.fetchMode = immutableConfig.getEnum(FETCH_MODE, FetchMode.NATIVE);
    }

    public void halt() {
        halted.set(true);
    }

    public void exitAndJoin() {
        halted.set(true);
        try {
            join();
        } catch (InterruptedException e) {
            LOG.error(e.toString());
        }
    }

    public boolean isHalted() {
        return halted.get();
    }

    @Override
    public void run() {
        Objects.requireNonNull(context);

        fetchMonitor.registerFetchThread(this);

        FetchItem fetchItem = null;

        try {
            while (!fetchMonitor.isMissionComplete() && !isHalted()) {
                fetchItem = schedule();

                if (fetchItem.getTask() == null) {
                    sleepAndRecord();
                    continue;
                }

                WebPage page = fetchOne(fetchItem);
                write(page.getKey(), page);

                ++taskCount;
            } // while
        } catch (final Throwable e) {
            LOG.error("Unexpected throwable : " + StringUtil.stringifyException(e));
        } finally {
            if (fetchItem != null && fetchItem.getTask() != null) {
                taskScheduler.finishUnchecked(fetchItem.getTask());
            }

            fetchMonitor.unregisterFetchThread(this);

            LOG.info("Thread #{} finished", getId());
        }
    }

    public void report() {
        if (servedHosts.isEmpty()) {
            return;
        }

        String report = String.format("Thread #%d served %d tasks for %d hosts : \n", getId(), taskCount, servedHosts.size());
        report += "\n";

        String availableHosts = servedHosts.stream()
                .map(PoolId::toUrl)
                .map(Urls::reverseHost).sorted().map(Urls::unreverseHost)
                .map(host -> String.format("%1$40s", host))
                .collect(Collectors.joining("\n"));

        report += availableHosts;
        report += "\n";

        REPORT_LOG.info(report);
    }

    private void sleepAndRecord() {
        fetchMonitor.registerIdleThread(this);

        try {
            Thread.sleep(1000);
        } catch (final Exception ignored) {
        }

        fetchMonitor.unregisterIdleThread(this);
    }

    private FetchItem schedule() {
        FetchJobForwardingResponse fetchResult = null;
        FetchTask fetchTask = null;

        if (fetchMode.equals(FetchMode.CROWDSOURCING)) {
            fetchResult = taskScheduler.pollFetchResult();

            if (fetchResult != null) {
                URL url = Urls.getURLOrNull(fetchResult.getQueueId());
                fetchTask = taskScheduler.getTasksMonitor().findPendingTask(fetchResult.getPriority(), url, fetchResult.getItemId());

                if (fetchTask == null) {
                    LOG.warn("Bad fetch item id {}-{}", fetchResult.getQueueId(), fetchResult.getItemId());
                }
            }
        } else {
            if (currQueueId == null) {
                fetchTask = taskScheduler.schedule();
            } else {
                fetchTask = taskScheduler.schedule(currQueueId);
            }

            if (fetchTask != null) {
                // the next time, we fetch items from the same queue as this time
                currQueueId = new PoolId(fetchTask.getPriority(), fetchTask.getProtocol(), fetchTask.getHost());
                servedHosts.add(currQueueId);
            } else {
                // The current queue is empty, fetch item from top queue the next time
                currQueueId = null;
            }
        }

        return new FetchItem(fetchTask, fetchResult);
    }

    private WebPage fetchOne(FetchItem fetchItem) {
        FetchTask task = fetchItem.getTask();

        WebPage page = fetchComponent.fetchContent(task.getPage());

        PoolId queueId = new PoolId(task.getPriority(), task.getProtocol(), task.getHost());
        taskScheduler.finish(queueId, task.getItemId());

        return page;
    }

    @SuppressWarnings("unchecked")
    private void write(String key, WebPage page) {
        try {
            context.write(key, page.unbox());
        } catch (IOException | InterruptedException e) {
            LOG.error("Failed to write to hdfs" + StringUtil.stringifyException(e));
        } catch (Throwable e) {
            LOG.error(StringUtil.stringifyException(e));
        }
    }

    @Override
    public int compareTo(@Nonnull FetchThread fetchThread) {
        return id - fetchThread.id;
    }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/crawl/fetch/FetchThread.kt
package ai.platon.pulsar.crawl.fetch

import ai.platon.pulsar.common.MetricsSystem
import ai.platon.pulsar.common.ReducerContext
import ai.platon.pulsar.common.StringUtil
import ai.platon.pulsar.common.Urls
import ai.platon.pulsar.common.config.ImmutableConfig
import ai.platon.pulsar.crawl.component.FetchComponent
import ai.platon.pulsar.crawl.fetch.data.PoolId
import ai.platon.pulsar.persist.RetryScope
import ai.platon.pulsar.persist.WebPage
import ai.platon.pulsar.persist.gora.generated.GWebPage
import ai.platon.pulsar.persist.metadata.FetchMode
import org.apache.hadoop.io.IntWritable
import org.slf4j.LoggerFactory
import java.io.IOException
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger
import kotlin.collections.HashMap

/**
 * This class picks items from queues and fetches the pages.
 */
class FetchThread(
        private val fetchMonitor: FetchMonitor,
        private val fetchComponent: FetchComponent,
        private val taskScheduler: TaskScheduler,
        immutableConfig: ImmutableConfig,
        private val context: ReducerContext<IntWritable, out IFetchEntry, String, GWebPage>
): Thread(), Comparable<FetchThread> {
    companion object {
        private val instanceSequence = AtomicInteger(0)
        val pendingTasks = ConcurrentHashMap<Int, WebPage>()
    }

    private val log = LoggerFactory.getLogger(FetchThread::class.java)

    private val id = instanceSequence.incrementAndGet()
    /**
     * Fix the thread to a specified queue as possible as we can
     */
    private var currPoolId: PoolId? = null
    private val servedPoolIds = TreeSet<PoolId>()
    private var taskCount = 0
    private val closed = AtomicBoolean(false)
    private val tasksMonitor = taskScheduler.tasksMonitor

    init {
        this.isDaemon = true
        this.name = "w$id" // w is short for worker
    }

    fun exit() {
        if (closed.compareAndSet(false, true)) {
            interrupt()
            join()
        }
    }

    override fun run() {
        fetchMonitor.registerFetchThread(this)


        var task: FetchTask? = null

        try {
            while (!fetchMonitor.isMissionComplete && !closed.get() && !interrupted()) {
                task = schedule()

                // TODO: block in schedule
                if (task == null) {
                    sleep()
                    continue
                }

                if (log.isTraceEnabled) {
                    log.trace("Ready to fetch task from pool {}", task.poolId)
                }

                val page = fetchComponent.fetchContent(task.page)
                taskScheduler.finish(task.poolId, task.itemId)

                if (page.protocolStatus.isRetry(RetryScope.PRIVACY_CONTEXT)) {
                    // 1. cancel all running tasks
                    // 2. push back to fetch queue
                    // 3. lock down fetch queue
                    // 4. reset fetch context
                    //
                    // tasksMonitor.produce(context.jobId, page)
                    pendingTasks[context.jobId] = page
                } else if (page.isNotNil) {
                    if (log.isInfoEnabled) {
                        log.info(MetricsSystem.getFetchCompleteReport(page))
                    }

                    write(page.key, page)
                }

                ++taskCount
            } // while
        } catch (e: Throwable) {
            log.error("Unexpected throwable : " + StringUtil.stringifyException(e))
        } finally {
            if (task != null) {
                taskScheduler.finishUnchecked(task)
            }

            fetchMonitor.unregisterFetchThread(this)

            log.info("Thread #{} finished", getId())
        }
    }

    fun report() {
        if (servedPoolIds.isEmpty()) {
            return
        }

        val report = StringBuilder()
        report.appendln(String.format("Thread #%d served %d tasks for %d hosts : \n", getId(), taskCount, servedPoolIds.size))

        servedPoolIds.map {
            Urls.reverseHost("${it.protocol}://${it.host}")
        }.sorted().map { Urls.unreverseHost(it) }.joinTo(report, "\n") { String.format("%1$40s", it) }
        report.appendln()

        log.info(report.toString())
    }

    private fun waitUntilAllFree() {

    }

    private fun sleep() {
        fetchMonitor.registerIdleThread(this)

        try {
            TimeUnit.SECONDS.sleep(1)
        } catch (ignored: InterruptedException) {}

        fetchMonitor.unregisterIdleThread(this)
    }

    private fun schedule(): FetchTask? {
        var fetchTask: FetchTask? = null

        val fetchMode = fetchMonitor.options.fetchMode
        if (fetchMode == FetchMode.CROWD_SOURCING) {
            val response = taskScheduler.pollForwardResponse()

            if (response != null) {
                val url = Urls.getURLOrNull(response.queueId)
                if (url != null) {
                    fetchTask = taskScheduler.tasksMonitor.findPendingTask(response.priority, url, response.itemId)
                }

                if (fetchTask == null) {
                    log.warn("Bad fetch item id {}-{}", response.queueId, response.itemId)
                }
            }
        } else {
            fetchTask = if (currPoolId == null) {
                taskScheduler.schedule()
            } else {
                taskScheduler.schedule(currPoolId)
            }

            if (fetchTask != null) {
                val poolId = fetchTask.poolId
                servedPoolIds.add(poolId)

                // the next time, we fetch items from the same queue as this time
                currPoolId = poolId
            } else {
                // The current queue is empty, fetch item from top queue the next time
                currPoolId = null
            }
        }

        if (fetchTask != null) {
            if (fetchMode != fetchTask.page.fetchMode) {
                log.warn("FetchTask.page.fetchMode {} is not expected {}", fetchMode, fetchTask.page.fetchMode)
            }
        }

        return fetchTask
    }

    private fun write(key: String, page: WebPage) {
        try {
            // the page is fetched and status are updated, write to the file system
            context.write(key, page.unbox())
        } catch (e: IOException) {
            log.error("Failed to write to hdfs - {}", StringUtil.stringifyException(e))
        } catch (e: InterruptedException) {
            log.error("Interrupted - {}", StringUtil.stringifyException(e))
        } catch (e: Throwable) {
            log.error(StringUtil.stringifyException(e))
        }
    }

    override fun compareTo(other: FetchThread): Int {
        return id - other.id
    }
}
